add Fedora dockerfile

TomasTomecek opened the pull request:
TODO:

* [x] update root README
* [ ] ~~update 5/README~~

notroj commented:
My assumptions (but really need @hhorak):

- we need to carry on assuming this will be usable as non-root user as in OpenShift
- so yes we should chmod 777 /etc/varnish as we do for the SCL equivalent paths
- in general we should parameterize everything, set env vars in Dockerfile to appropriate paths for the packages used, and have scripts not dependent on SCL or non-SCL specific package paths

omron93 commented:
> so yes we should chmod 777 /etc/varnish as we do for the SCL equivalent paths

`777` are too open. Use `fix-permissions` [1] which sets group of files/dirs to `root` and write permissions to this group. User 1001 used in this images is in `root` group and this will solve permissions problem when running images with random UID in OpenShift (they are using group 0).

[1[ https://github.com/sclorg/s2i-base-container/blob/master/core/root/usr/bin/fix-permissions

TomasTomecek commented:
@omron93 @notroj thanks for your suggestions!

Is there an up-to-date, Fedora s2i base image available somewhere?

TomasTomecek commented:
@omron93 please take a look: tests are now passing for me locally. Once you approve the code changes, I'll update READMEs.

omron93 commented:
[test]

TomasTomecek commented:
rebased: the gcc fix should be picked up

omron93 commented:
[test]

omron93 commented on test/run:
```diff
@@ -162,7 +167,11 @@ test_application() {
+  if CONTAINER_ARGS="--rm" container_run rpm -q scl-utils; then
+    test_scl_usage "varnishd -V" "varnish-4"
+  else
+    test_scl_usage "varnishd -V" "varnishd (varnish-"
```
Would be great to test right `varnishd` versions for Fedora too...

You can use `VERSION` variable which is set to name of directory (~SW version).

`test_scl_usage "varnishd -V" "varnishd (varnish-$VERSION"`
IMHO this could be used for both - SCL and Fedora images...

TomasTomecek replied:
good suggestion, will do!

omron93 commented on s2i/bin/run:
```diff
@@ -0,0 +1,31 @@
+
+. ${VARNISH_CONFIGURATION_PATH}/varnish.params
+
+if grep "NAME=Fedora" /etc/os-release; then
```
Why Fedora differs?
(I don't know varnish much, so that is why I'm asking)

TomasTomecek replied:
(fyi, me neither)

The CLI options of varnishd were changed significantly between version 4 and 5. No backwards compatibility. Pretty nuts.

omron93 replied:
Thanks for explanation

omron93 replied:
I was thinking about this more.

Varnish 4.1 changed that parameters which differs. It is not about Fedora but it will probably change even for newer varnish in some SCL. So it would be better to base the params on varnishd version.

TomasTomecek commented:
updated

omron93 commented:
[test]

omron93 commented on s2i/bin/run:
```diff
@@ -0,0 +1,31 @@
+	exec varnishd \
+		-F \
+		-f $VARNISH_VCL_CONF \
+		-a :8080 \
```
To not to change behavior `-a :8080 -a :8443` should be used.

omron93 commented on s2i/bin/run:
```diff
@@ -0,0 +1,31 @@
+		-a :8080,:8443 \
+		-T 127.0.0.1:8081 \
+		-t $VARNISH_TTL \
+		-u default \
```
`-j "unix,user=default,ccgroup=root"` can be used for newer varnishd.

So IMPOV it would be more readable to have if/else for varnish version -> set right parameters there (`-a X,Y` vs `-a X -a Y` or `-u`,`-g` vs `-j XY`). And then have only one `exec varnishd ...`

TomasTomecek replied:
Hmmm, docs seem to be misleading:

```
[INFO] Waiting for application container to start...

Can't read /etc/scl/conf/rh-varnish4, rh-varnish4 is probably not installed.
Error: Too many arguments (unix,user=varnish,ccgroup=root...)
(-? gives usage)
```

TomasTomecek replied:
(never mind, I'm a moron, the problem was elsewhere)

omron93 commented on 5/Dockerfile.fedora:
```diff
@@ -0,0 +1,39 @@
+      io.openshift.expose-services="8080:http,8443:https" \
+      io.openshift.tags="builder,varnish" \
+      name="f27/varnish" \
+      version="0"
```
According to Fedora guidelines `release`, `architecture`, `maintainer` and `usage/run` labels are missing. Having right spec in github will help when someone will try to get the image into Fedora (pass container review).

Is inhering them from base image enough?

TomasTomecek replied:
You're right! I'm adding release, maintainer and usage; architecture should be inherited.

omron93 commented on 5/Dockerfile.fedora:
```diff
@@ -0,0 +1,39 @@
+      io.k8s.display-name="Varnish 5" \
+      io.openshift.expose-services="8080:http,8443:https" \
+      io.openshift.tags="builder,varnish" \
+      name="f27/varnish" \
```
Better to use `$FGC` instead of `f27`. (same in s2i `usage` script)

omron93 commented on test/run:
```diff
@@ -35,6 +35,11 @@ run_s2i_build() {
   s2i build ${s2i_args} file://${test_dir}/test-app ${IMAGE_NAME} ${IMAGE_NAME}-testapp
 }
 
+container_run() {
```
It is not used anywhere.

TomasTomecek commented:
Thanks for such a thorough review. I really appreciate it.

I won't be able to finish the PR this year likely.

TomasTomecek commented:
I addressed all the comments. The tests are passing for me locally (centos & version 4, fedora & version 5).

@omron93 please let me know if there is anything I should fix.

hhorak commented:
[test]

omron93 commented:
LGTM.

I'm still not able to build s2i-core in Fedora:-( @TomasTomecek @hhorak Could you try this please?

TomasTomecek commented:
I can definitely try.

TomasTomecek commented:
https://pagure.io/releng/issue/7230

hhorak commented:
I wouldn't block merging this just because of Fedora infra, the Fedora image should work fine when rebuilding locally, so it's valuable even with those issues. Merging.
